### PR TITLE
Add a generic install.sh and update README.md.

### PR DESCRIPTION
**Pull request by kwlzn**
* Adds a generic `install.sh` utility.
* Updates `README.md` to use the new installer.
* Adds tests.

OSX install example:

```
om:~ kw$ curl -LSsf https://raw.githubusercontent.com/kwlzn/lift/kwlzn/installer/install.sh | /bin/bash
Download URL is: https://github.com/a-scie/lift/releases/latest/download/science-fat-macos-aarch64
######################################################################## 100.0%
######################################################################## 100.0%
Download completed successfully
Download matched it's expected sha256 fingerprint, proceeding
Installed https://github.com/a-scie/lift/releases/latest/download/science-fat-macos-aarch64 to /Users/kw/.local/bin/science
WARNING: /Users/kw/.local/bin is not detected on $PATH
You'll either need to invoke /Users/kw/.local/bin/science explicitly or else add /Users/kw/.local/bin to your shell's PATH.
om:~ kw$ /Users/kw/.local/bin/science -V
0.7.0
```

### CODE REVIEW

File: install.sh
Comment (jsirois) [inline install.sh:47]: No need for all this, in addition to `https://github.com/a-scie/lift/releases/download/v0.7.0/...` they support `https://github.com/a-scie/lift/releases/latest/download/...`.

File: install.sh
Comment (jsirois) [inline install.sh:59]: This choice, though mine as well, will anger some folks. XDG calls out `~/.local/bin`: https://www.freedesktop.org/software/systemd/man/latest/file-hierarchy.html#Home%20Directory

File: install.sh
Comment (jsirois) [inline install.sh:34]: That's disingenuous since we release for Windows! It would be more honest to say not supported, download from here, etc. at the least.

File: install.sh
Comment (jsirois) [inline install.sh:24]: I think you want amd64 as well to cover the bases.

**Review by jsirois — COMMENTED** [review]
Thanks for chipping in.

I had left things as advisory - aka I leave it to you to know how to download a thing and check a hash and not tackled the bigger problem of an installer until 1.0. Getting a head start on that is good, but I'd definitely want these things in the end:
+ science.scie.app docs have an install page with instructions / links and the instructions should probably use a sice.app URL (if there were install scripts and they lived in this repo, they could be symlinked into the doc tree so that they get published to the doc site).
+ Install should be at least explained for each supported platform, ideally though, it would be supported via script or officially supported package.

There is roughly that for scie-pants for example between:
+ https://www.pantsbuild.org/2.21/docs/getting-started/installing-pants
+ https://github.com/pantsbuild/setup/blob/gh-pages/get-pants.sh (which works on Windows Git-supplied bash in addition to linux and mac)
+ https://github.com/pantsbuild/scie-pants/blob/d50bd33fd67e944384c9548811a66c1cb03113a5/.github/workflows/release.yml#L156-L178

I'm not sure what you were gunning for here @kwlzn. A step towards that sort of thing maybe?

File: install.sh
Comment (jsirois) [reply install.sh:59]: (scie-pants got this complaint FWIW and switched)

**Comment by kwlzn** [comment]
> I'm not sure what you were gunning for here @kwlzn. A step towards that sort of thing maybe?

yeah, a pre-1.0 rough edged step in that direction. essentially, just aiming to make it trivial to link to and have a complete newcomer running `science -h` (as docs) in <1 minute in case it comes up in integration discussions with other OSS projects.

I'm happy to refine this a bit more tho and port `get_pants.sh`'s approach/logic + add some CI coverage with 1.0 in mind.

File: install.sh
Comment (kwlzn) [reply install.sh:47]: I thought so too, but didn't work for me on an initial test so I punted - will revert.

File: install.sh
Comment (kwlzn) [reply install.sh:59]: sgtm

File: install.sh
Comment (kwlzn) [reply install.sh:34]: punted on windows because supporting it is unfamiliar in the context of UNIX shell script execution to me (I'd assume a windows-mode installer would be e.g. `install.bat` or `install.ps1`), but I suppose the windows mode of support here is just "run in WSL" which behaves largely as UNIX.

File: install.sh
Comment (kwlzn) [reply install.sh:24]: rgr

File: install.sh
Comment (jsirois) [inline install.sh:71]: No need for an extra dep, `arch` == `uname -m` and you use `uname -s` above.

File: install.sh
Comment (jsirois) [inline install.sh:110]: You use dirname as well below and its not ensured yet.

File: README.md
Comment (jsirois) [inline README.md:34]: Now `~/.local/bin/science -V` by default ... but maybe not needed aince you now get the warning if science is not on the PATH.

File: tests/test_installer.py
Comment (jsirois) [inline tests/test_installer.py:8]: This will work fine, but FYI there is a build_root Path fixture in conftest.

File: tests/test_installer.py
Comment (jsirois) [inline tests/test_installer.py:13]: This is Python 3.12 so you can omit the encoding if you want, use `:=` match cases, etc.

File: install.sh
Comment (jsirois) [reply install.sh]: > but I suppose the windows mode of support here is just "run in WSL" which behaves largely as UNIX.

Nope. So, the reasoning goes like this:
1. If you use science, you are, by definition, a developer
2. Developers will almost always use either git or wsl or they'll have wsl installed for Docker use since its the most performant Docker backend on Windows.

Item 2 ensures bash. Git installs its own MSYS2 bash to even be able to work. Windows does likewise when it has wsl installed.

So you can almost always run bash from real Windows (not wsl) when you use Windows and are a developer. The scie-jump ITs, for example, run on Windows in CI and those are powered by an examples/run.sh bash script.
--
All that said, there is no excuse. I need to learn PowerShell and come up with a 1st class script. They release for Linux, Mac, Windows, so really no excuse: https://github.com/PowerShell/PowerShell/releases/tag/v7.4.5

**Review by jsirois — APPROVED** [review]
LGTM, we'll see what CI says.

File: tests/test_installer.py
Comment (jsirois) [inline tests/test_installer.py:20]: This test fails on the macos-12 (x86-64) shard as does the one below in the same check. Since the output is swallowed in both cases there is not much to go on currently. You'll have to experiment.

File: install.sh
Comment (kwlzn) [reply install.sh:71]: sgtm

File: install.sh
Comment (kwlzn) [reply install.sh:110]: fixed

File: README.md
Comment (kwlzn) [reply README.md:34]: fixed

File: tests/test_installer.py
Comment (kwlzn) [reply tests/test_installer.py:8]: thanks, moved to that

File: tests/test_installer.py
Comment (kwlzn) [reply tests/test_installer.py:20]: appears from this debug log output (which also catches an error-mode print bug) to be related to output of `arch` vs `uname`, : https://github.com/a-scie/lift/actions/runs/10721551555/job/29749027884?pr=88#step:9:256

on my x86 mac, `arch` says `i386` and `uname -m` says `x86_64`.

File: install.sh
Comment (kwlzn) [reply install.sh]: aha, but why port any logic to powershell interpreted script if powershell can just run bash natively?

File: tests/test_installer.py
Comment (kwlzn) [reply tests/test_installer.py]: ah neat, modernized it a bit but the opportunity to use match cases eluded me - let me know if you see something I don't.

File: tests/test_installer.py
Comment (jsirois) [inline tests/test_installer.py:17]: This is failing on Windows because Windows does not do shebangs. The scie-jump tests don't have this problem because the script is run directly by the job step (which runs in bash): https://github.com/a-scie/jump/blob/83bfd085d428f5c0746acdab45d7bb401aba28f8/.github/workflows/ci.yml#L58

I think you'll need args: `["bash", script, ...]`.

File: install.sh
Comment (jsirois) [reply install.sh]: Curl did not work. I can't remember the details, but after alot of fail, using pwsh to do downloads worked.

File: tests/test_installer.py
Comment (jsirois) [reply tests/test_installer.py:17]: Sent you an invite so you can self-serve run /  retry CI.

Comment (kwlzn) [reply tests/test_installer.py:20]: 👍

Comment (jsirois) [reply tests/test_installer.py:20]: Yeah, so you're getting the wsl bash via PATH, it looks like you need the git bash. I think plumbing an env var from the CI step that captures the current active bash the runner step is using (which is the right bash) is needed.

File: install.sh
Comment (jsirois) [inline install.sh:198]: Aha - Ok. Yeah, just a bad URL for the Windows case. Should be good to go after that fix.

File: tests/test_installer.py
Comment (jsirois) [inline tests/test_installer.py:16]: This obviously works with the GH runners but I'll probably circle back and either plumb the bash executable path / specialize in CI or else just implement the PowerShell equivalent 1st class. I looked at examples from uv, deno and bun and there is not too much PowerShell to learn.

**Review by jsirois — APPROVED** [review]
Thanks for working on this Kris.

File: install.sh
Comment (jsirois) [reply install.sh:195]: You took my meaning wrong. You're missing adding an .exe on the end of this URL for Windows. There _are_ Windows releases for both x86-64 and aarch64, search for ".exe" here: https://github.com/a-scie/lift/releases/tag/v0.7.0

You'll need to bump you 0.6.1 to 0.7.0 though so that the tests work on Windows arm64.

File: tests/test_installer.py
Comment (kwlzn) [reply tests/test_installer.py]: hmm, I think just checking if the git-bash path exists first - then if not falling back to "bash.exe" (which did work initially, but gave an error about WSL not having the modules installed to run bash etc) - should provide unilateral support for the "windows developer that has git bash and/or WSL installed" env assumption?

either way, feel free to iterate accordingly - esp if there are broader cases of a need for discovering a usable bash for testing etc.

File: tests/test_installer.py
Comment (jsirois) [inline tests/test_installer.py:22]: You took my meaning wrong here too. `C:\` is not god given, a user can install on `D:\` or wherever. I'm fine with what you had prior. Like I said, I'll follow-up.

File: install.sh
Comment (kwlzn) [reply install.sh:195]: aha, I was looking at https://github.com/a-scie/lift/releases and the lack of x86_64 (need to click "show all") was a red herring. thanks!

File: tests/test_installer.py
Comment (jsirois) [reply tests/test_installer.py]: Yeah, you mistook my meaning. I just mean to do something like this in the CI workflow step that runs tests:
```yaml
run: |
  export _PEX_TEST_USE_BASH=$SHELL
  ... rest ...
```

Then in the test, use that env var if set.

Comment (jsirois) [reply tests/test_installer.py]: Basically, by trying to do this test using pytest, you inherit the extra burden of having to simulate the user env. vs just running the install.sh script via CI bash / bash tests.

File: install.sh
Comment (jsirois) [inline install.sh:195]: Ooops!

File: tests/test_installer.py
Comment (kwlzn) [reply tests/test_installer.py]: aha, and I suppose its not as simple as just plugging in e.g. `%ProgramFiles%\Git\bin\bash.EXE`?

windows is a foreign land to me, but my base assumption is/was that most people don't tend to tamper with the default - and thus good enough to hardcode.

File: install.sh
Comment (kwlzn) [reply install.sh]: (just testing the windows case locally - accidentally pushed without revert)

File: tests/test_installer.py
Comment (jsirois) [reply tests/test_installer.py]: > windows is a foreign land to me, but my base assumption is/was that most people don't tend to tamper with the default 

I agree, buts devs are the audience here and they are exactly the people that screw with things.

File: tests/test_installer.py
Comment (kwlzn) [reply tests/test_installer.py:16]: gotcha. I was definitely thinking about both CI and developers running these tests on Windows.

**Comment by kwlzn** [comment]
the salient part of the latest CI failure seems to be a failure in the pwsh curl mode call:

```
pwsh -c \'Invoke-WebRequest -OutFile /c/actions-runner/_work/lift/lift/.nox/f83bbc49/tmp/tmp.LGZBCkLskN/science-fat-windows-x86_64.exe.sha256 -Uri https://github.com/a-scie/lift/releases/download/v0.7.0/science-fat-windows-x86_64.exe.sha256\'
 Invoke-WebRequest: 
 Could not find a part of the path \'C:\\c\\actions-runner\\_work\\lift\\lift\\.nox\\f83bbc49\\tmp\\tmp.LGZBCkLskN\\science-fat-windows-x86_64.exe.sha256\'.
```

my intuition says the `/c/...` in `/c/actions-runner/...` implies the `C:\` bit in alternate form, but pwsh is prepending to that to arrive at `C:\c\...`. need to do some more digging.

**Comment by jsirois** [comment]
I think probably not. It seems highly likely dirname / basename don't play well with \.

**Comment by kwlzn** [comment]
> I think probably not. It seems highly likely dirname / basename don't play well with \.

afaict, `basename` is used exactly once, with a URL and not a filepath, so that's natively using `/` always and spits out a single path part sans any path separators. and `dirname` is only ever called on OSX as a x-platform way to net `install -D` behavior. almost all of the logic leading up to this point is lifted directly from `get-pants.sh`, which presumably is known to work on Windows - which is the most baffling part.

I did notice one mixed `\` + `/` path construction in the script in debug logging (due to `-d` receiving a windows temp dir path from the `TempPathFactory` fixture), but already fixed that in [baba4f4](https://github.com/a-scie/lift/pull/88/commits/baba4f4d6fb27057437e5a6170d8b6216354a6f5).

there is one other similar path separator join (directly adjacent to where `basename` is used), but that's only ever joined with a `mktemp -d` output which appears to use `/` exclusively on windows fwict.

**Comment by kwlzn** [comment]
explicitly converting the `/c/x/y/z` `mktemp`-returned MSYS POSIX format to `C:\x\y\z` via `cygpath -w` in https://github.com/a-scie/lift/pull/88/commits/5c1a74e6c91c078c54a1b8124fe7f6494d09d13f definitely got past the initial pwsh iwr failure fwiw.

but based on https://devblogs.microsoft.com/commandline/tar-and-curl-come-to-windows/ it seems windows does have native curl since windows 10 (2018-present - not sure if support needs to go back further than that?) - it definitely works as expected at least in CI, so I moved to that as a unification.

with the broken downloads fixed, we're now onto a shasum failure:

```
sha256sum: \'science-fat-windows-x86_64.exe\'$\'\\r\\r\': No such file or directory
```

which seems to indicate some trailing carriage returns that may be messing things up:

```
om:tmp kw$ cat -e science-fat-windows-x86_64.exe.sha256 
010cfa5ddf8d50482dada3624c044d74dc8e699d7b218d3d8d74a89204a4029c *science-fat-windows-x86_64.exe^M^M$
```

(with shasum checking removed, the tests pass E2E on the windows shards)

**Comment by kwlzn** [comment]
sanitizing the downloaded sha256 file with `tr` on windows seems to have done the trick and all tests now pass.

so, I think this is good to go - assuming you're ok with the switch to `curl`. there appear to be other failures from `test_exe.py` on the windows shard tho. is that expected?

File: install.sh
Comment (jsirois) [inline install.sh:124]: Hrm, my bug: https://github.com/a-scie/lift/blob/bf8edd195cace2916ae76271c1330ba3c6da2a45/science/commands/build.py#L81
2 bugs really, using `\r\n` when science is run on Windows, and using the host OS linesep for a shasum file for a potentially foreign scie. Afaict after some brief research, shasum files are not a thing in Windows; so I can just always use a `\n` linesep.

Your fix needs to remain to handle old releases. but I'll fix that.

**Review by jsirois — APPROVED** [review]
The remaining wonkiness was just self hosted runner hygiene. I've fixed that up with a custom runner job start / stop script and I'll follow up with a few CI workflow tweaks as well towards that end.

**Comment by jsirois** [comment]
Thanks for wading through this Kris.

> so, I think this is good to go - assuming you're ok with the switch to `curl`.

Absolutely - simpler. I was forced to pwsh 2 years ago when getting scie-jump green in CI. I'm not sure what is different this much time passed.

File: install.sh
Comment (jsirois) [reply install.sh:124]: Fixed in #92

**Pull request merged**
